docs: add new website using Astro Starlight (#5246)
mochajs/mocha: docs-next/src/content/docs/running/cli.mdx

---
title: Command-Line Usage
description: Running Mocha in your terminal.
---

> Running `npx mocha --help` will show the full list of available commands:

Reviewer: Should we add a source comment linking this to the definition in code? I want to update some stuff here but want it to match released code too

Reviewer: Good candidate for a followup PR

```plaintext
mocha [spec..]

Run tests with Mocha

Commands
  mocha inspect [spec..]  Run tests with Mocha                         [default]
  mocha init <path>       create a client-side Mocha setup at <path>

Rules & Behavior
      --allow-uncaught              Allow uncaught errors to propagate [boolean]
  -A, --async-only                  Require all tests to use a callback (async)
                                    or return a Promise                [boolean]
  -b, --bail                        Abort ("bail") after first test failure
                                                                       [boolean]
      --check-leaks                 Check for global variable leaks    [boolean]
      --delay                       Delay initial execution of root suite
                                                                       [boolean]
      --dry-run                     Report tests without executing them[boolean]
      --exit                        Force Mocha to quit after tests complete
                                                                       [boolean]

>       --pass-on-failing-test-suite  Not fail test run if tests were failed

Reviewer: ```suggestion
      --pass-on-failing-test-suite  Don't fail test run if tests were failed
```

Reviewer: Mark as resolved, just proofreading

### `--bail, -b`

Causes Mocha to stop running tests after the first test failure it encounters.
Corresponding "after each" and "after all" hooks are executed for potential cleanup.

`--bail` does _not_ imply `--exit`.

### `--check-leaks`

Use this option to have Mocha check for global variables that are leaked while running tests.
Specify globals that are acceptable via the `--global` option (for example: `--check-leaks --global jQuery --global MyLib`).

### `--compilers`

:::note
`--compilers` was removed in v6.0.0.
See [further explanation and workarounds](https://github.com/mochajs/mocha/wiki/compilers-deprecation).
:::

### `--delay`

Delay initial execution of root suite.

See [Delayed Root Suite](../features/hooks#delayed-root-suite).

### `--dry-run`

:::note[New in v9.0.0]
:::

Report tests without executing any of them, neither tests nor hooks.

### `--exit`

:::note[Updated in v4.0.0]
TL;DR: If your tests hang after an upgrade to Mocha v4.0.0 or newer, use `--exit` for a quick (though not necessarily recommended) fix.
:::

_Prior to_ version v4.0.0, _by default_, Mocha would force its own process to exit once it was finished executing all tests.

> This behavior enables a set of potential problems; it's indicative of tests (or fixtures, harnesses, code under test, etc.) which don't clean up after themselves properly.

Reviewer: ```suggestion
This behavior enables a set of potential problems: it's indicative of tests (or fixtures, harnesses, code under test, etc.) which don't clean up after themselves properly.
```

Reviewer: Mark as resolved, just proofreading

It can also be something like a runaway `setInterval()`, or even an errant `Promise` that never fulfilled.

The _default behavior_ in v4.0.0 (and newer) is `--no-exit`, where previously it was `--exit`.

**The easiest way to "fix" the issue is to pass `--exit` to the Mocha process.**
It _can_ be time-consuming to debug &mdash; because it's not always obvious where the problem is &mdash; but it _is_ recommended to do so.

To ensure your tests aren't leaving messes around, here are some ideas to get started:

- See the [Node.js guide to debugging](https://nodejs.org/en/docs/inspector)
- Use the new [`async_hooks`](https://github.com/nodejs/node/blob/master/doc/api/async_hooks.md) API ([example](https://gist.github.com/boneskull/7fe75b63d613fa940db7ec990a5f5843))
- Try something like [wtfnode](https://npm.im/wtfnode)
- Use [`.only`](../declaring/exclusive-tests) until you find the test that causes Mocha to hang

### `--pass-on-failing-test-suite`

:::note[New in v10.7.0]
:::

If set to `true`, Mocha returns exit code `0` even if there are failed tests during the run.

### `--fail-zero`

:::note[New in v9.1.0]
:::

> Fail test run if no tests are encountered with `exit-code: 1`.

Reviewer: Previous reading could be that a test needs an exit code

```suggestion
Fail test run with `exit-code: 1` if no tests are encountered.
```

Reviewer: Mark as resolved, just proofreading

  See [Babel](https://github.com/mochajs/mocha-examples/tree/main/packages/babel) or [TypeScript](https://github.com/mochajs/mocha-examples/tree/main/packages/typescript) working examples.

Modules required in this manner are expected to do work synchronously; Mocha won't wait for async tasks in a required module to finish.

:::caution
**You cannot use `--require` to set hooks**.
If you want to set hooks to run, e.g., before each test, use a [Root Hook Plugin](../features/root-hook-plugins).
:::

:::note
As of v8.0.0, Mocha supports `--require` for [NodeJS native ESM](../explainers/nodejs-native-esm-support).
There is no separate `--import` flag.
:::

### `--sort, -S`

:::caution
`--sort` is incompatible with [parallel mode](../features/parallel-mode).
:::

Sort test files (by absolute path) using [Array.prototype.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort).

### `--watch, -w`

Rerun tests on file changes.

The `--watch-files` and `--watch-ignore` options can be used to control which files are watched for changes.

> Tests may be rerun manually by typing &#x24e1; &#x24e2; &#x23ce; (same shortcut as `nodemon`).

Reviewer: Why `ⓡ ⓢ ⏎` ?

```suggestion
Tests may be rerun manually by typing `rs` and pressing enter (same shortcut as `nodemon`).
```

Reviewer: Mark as resolved, just proofreading

The option can be given multiple times.
The option accepts a comma-delimited list: `--watch-ignore a,b` is equivalent to `--watch-ignore a --watch-ignore b`.

## Test Filters

### `--fgrep <string>, -f <string>`

:::caution[Breaking change in v6.0.0]
Now mutually exclusive with `--grep`.
:::

Cause Mocha to only run tests having titles containing the given `string`.

Mutually exclusive with `--grep`.

### `--grep <regexp>, -g <regexp>`

:::caution[Breaking change in v6.0.0]
Now mutually exclusive with `--fgrep`.
:::

Cause Mocha to only run tests matching the given `regexp`, which is internally compiled to a [RegExp](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Regexp).

Suppose, for example, you have "api" related tests, as well as "app" related tests, as shown in the following snippet; One could use `--grep api` or `--grep app` to run one or the other.
The same goes for any other part of a suite or test-case title, `--grep users` would be valid as well, or even `--grep GET`.

And another option with double quotes: `--grep "groupA|groupB"`.

And for more complex criteria: `--grep "/get/i"`.

> Some shells as e.g. Git-Bash-for-Windows may require: `--grep "'/get/i'"`.

Reviewer: ```suggestion
Some shells (e.g. Git Bash for Windows) may require: `--grep "'/get/i'"`.
```

Reviewer: Mark as resolved, just proofreading

### `--invert`

Use the _inverse_ of the match specified by `--grep` or `fgrep`.

Requires either `--grep` or `--fgrep` (but not both).

## Other Options

### `--inspect, --inspect-brk, inspect`

Enables Node.js' inspector.

Use `--inspect` / `--inspect-brk` to launch the V8 inspector for use with Chrome Dev Tools.

Use `inspect` to launch Node.js' internal debugger.

All of these options are mutually exclusive.

Implies `--timeout 0`.

## About Option Types

:::note[Updated in v6.0.0]
:::

Each flag annotated of type `[boolean]` in Mocha's `--help` output can be _negated_ by prepending `--no-` to the flag name.
For example, `--no-color` will disable Mocha's color output, which is enabled by default.

Unless otherwise noted, _all_ boolean flags default to `false`.

## About `node` Flags

The `mocha` executable supports all applicable flags which the `node` executable supports.

These flags vary depending on your version of Node.js.

`node` flags can be defined in Mocha's [configuration](./configuring).

:::note[New in v9.1.0]
You can also pass `node` flags to Node.js using [`--node-option`](#--node-option-name--n-name).
:::

### `--enable-source-maps`

> :::note[New in vNode.js v12.12.0]

Reviewer: ```suggestion
:::note[New in Node.js v12.12.0]
```

Reviewer: Mark as resolved, just proofreading

:::

If the [`--enable-source-maps`](https://nodejs.org/dist/latest-v12.x/docs/api/cli.html#cli_enable_source_maps) flag
is passed to Mocha, source maps will be collected and used to provide accurate stack traces for transpiled code:

```bash
Error: cool
    at Object.<anonymous> (/Users/fake-user/bigco/nodejs-tasks/build/src/index.js:27:7)
        -> /Users/fake-user/bigco/nodejs-tasks/src/index.ts:24:7
```

## About V8 Flags

Prepend `--v8-` to any flag listed in the output of `node --v8-options` (excluding `--v8-options` itself) to use it.

> V8 flags can be defined in Mocha's [configuration](./configuring).

Reviewer: I thought this was Mocha v8 at first, was confused, a link might help. Couldn't find a link to an official flags list

```suggestion
[V8](https://v8.dev/) flags can be defined in Mocha's [configuration](./configuring).
```

Reviewer: Mark as resolved, just proofreading
